### ISSUE TITLE
async suspend generates invalid LLVM within optional blocks

### ISSUE THREAD

**Issue by fengb** [ziglang/zig#4065]
```zig
pub fn unroll(comptime opt: ?usize) void {
    if (opt) |val| {
        var i: usize = 0;
        // -- broken LLVM module found: Instruction does not dominate all uses!
        // Using any of these alternatives fixes the problem:
        //
        // while (i < opt.?) : (i += 1) {
        // inline while (i < val) : (i += 1) {
        while (i < val) : (i += 1) {
            suspend;
        }
    }
}

test "unroll" {
    _ = async unroll(null);
}
```

**Comment by fengb** [comment]
It looks like any non-trivial block triggers it, not just loops. I think it’s missing a spill for the block variable in general, but simple cases like sequential suspends (inline loops) is optimized away.

**Comment by fengb** [comment]
Minimal repro:

```zig
pub fn unroll(opt: ?usize) usize {
    if (opt) |val| {
        suspend;
        return val;
    }
    return 0;
}

test "unroll" {
    _ = async unroll(null);
}

**Comment by fengb** [comment]
Probably related — error unions:

```zig
pub fn err() !usize {
    return error.Shtahp;
}

pub fn payload() usize {
    if (err()) |val| {
        suspend;
        return val;
    } else |err| {
        return 0;
    }
}

// Interestingly this works fine
pub fn justFine() usize {
    if (err()) |val| {
        return val;
    } else |err| {
        suspend;
        return 0;
    }
}
```


Also, while loops are internally different but have the same bugs:

```zig
pub fn unroll(opt: ?usize) usize {
    while (opt) |val| {
        suspend;
        return val;
    }
    return 0;
}

pub fn err() !usize {
    return error.Shtahp;
}

pub fn payload() usize {
    while (err()) |val| {
        suspend;
        return val;
    } else |err| {
        return 0;
    }
}
```

**Comment by andrewrk** [comment]
the commit said

> Partially fixes #4065

**Comment by frmdstryr** [comment]
I updated today (from 5 days ago) and am getting this error now in code that was working before.  It works if I remove the optional block and just use `.?`.

The exact error is:

```
Code Generation [1/398] broken LLVM module found: Instruction does not dominate all uses!
  %73 = getelementptr inbounds %"?std.linked_list.Node", %"?std.linked_list.Node"* %70, i32 0, i32 0, !dbg !9179
  %143 = getelementptr inbounds %std.linked_list.Node, %std.linked_list.Node* %73, i32 0, i32 2, !dbg !9212
```

The last commit that works https://github.com/ziglang/zig/commit/d8965002593c111069862e7de06402ee77b2b614 from 4 days ago https://github.com/ziglang/zig/commits/master?after=ab46713fa6eb7603e604320b77a4d101d30a0998+69 which is before this was merged so I don't think it's due to this PR. 

It can be reproduced with:

```zig
const std = @import("std");
const assert = std.debug.assert;
const io_mode = .evented;

const TaskQueue = std.atomic.Queue(Task);

const Task = struct {
    name: []const u8 = "",
    done: bool = false,
};

const Manager = struct {
    task: ?TaskQueue.Node = null,

    pub fn doTask(self: *Manager) void {
        self.task = TaskQueue.Node{
            .next = undefined,
            .prev = undefined,
            .data = Task{.name="Test"},
        };
        // actual code puts it in the queue
        if (self.task) |*t| {
            loop.yield();
            t.data.done = true;
        }
    }
};

pub fn main() void {
    var mgr = Manager{};
    mgr.doTask();
}

pub const Loop = struct {
    frame: ?anyframe = null,
    pub fn run(self: *Loop) void {
        var f = async main();
        while (true) {
            if (self.frame) |frame| {
                resume frame;
            }
        }
    }

    pub fn yield(self: *Loop) void {
        suspend {
            self.frame = @frame();
        }
        self.frame = null;
    }
};

var loop = Loop{};

export fn ResetHandler() void {
    loop.run();
}

```

Using the freestanding os.

**Comment by andrewrk** [comment]
Fixed by #4404

**Comment by andrewrk** [comment]
Landed in 014f66e6de4aaf81f32c796b12f981326a479397
